client/lib/perfmon/index.js

```diff
@@ -0,0 +1,178 @@
+
+	// check if there are any placeholders on the screen,
+	// and trigger a timing event when all the placeholders are gone or the user
+	// has navigated
+	checkForVisiblePlaceholders: function( trigger ) {
+
+		// determine how many placeholders are active in the viewport
+		var visibleCount = 0;
+		activePlaceholders.forEach( function( node ) {
+			if ( this.isElementVisibleInViewport( node ) ) {
+				visibleCount += 1;
+			}
+		}.bind( this ) );
+
+		// record event and reset timer if all placeholders are loaded OR user has just navigated
+		if ( placeholdersVisibleStart && ( visibleCount === 0 || trigger === 'navigate' ) ) {
+			// tell tracks to record duration
+			var duration = Date.now() - placeholdersVisibleStart;
+			debug(`Recording placeholder wait. Duration: ${duration}, Trigger: ${trigger}`);
+			analytics.pageLoading.record( `placeholder-wait-${trigger}`, duration );
+			placeholdersVisibleStart = null;
+		}
+
+		// if we can see placeholders, placeholdersVisibleStart is falsy, start the clock
+		if ( visibleCount > 0 && !placeholdersVisibleStart ) {
+			placeholdersVisibleStart = Date.now(); // TODO: performance.now()?
+		}
+
+		// if there are placeholders hanging around, print some useful stats
+		if ( activePlaceholders.length > 0 ) {
+			debug("Active placeholders: "+activePlaceholders.length);
+			debug("Visible in viewport: "+visibleCount);
+		}
+	},
+
+	isPlaceholder: function( node ) {
+		var className = node.className;
+		return className && className.indexOf && PLACEHOLDER_CLASSES.some( function( clazz ) { 
+			return (className.indexOf( clazz ) >= 0); 
+		} );
+	},
+
+	// adapted from http://stackoverflow.com/questions/123999/how-to-tell-if-a-dom-element-is-visible-in-the-current-viewport/7557433#7557433
+	isElementVisibleInViewport: function( node ) {
+		var rect = node.getBoundingClientRect();
+
+		// discard if width or height are zero
+		if ( (rect.top - rect.bottom) === 0 || (rect.right - rect.left) === 0) {
+			return false;
+		}
+
+		return (
+			rect.top >= 0 &&
+			rect.left >= 0 &&
+			rect.bottom <= (window.innerHeight || document.documentElement.clientHeight) && /*or $(window).height() */
+			rect.right <= (window.innerWidth || document.documentElement.clientWidth) /*or $(window).width() */
+		);
+	},
+
+	recordPlaceholders: function( mutation ) { 
+		var nodes = [];
+
+		if ( mutation.attributeName === 'class' ) { // mutation.type === 'attributes' is redundant
+			nodes = [mutation.target];
+		} else if ( mutation.type === 'childList' && mutation.addedNodes.length > 0 ) {
+			nodes = mutation.addedNodes;
+		} else {
+			return;
+		}
+
+		each( nodes, function( node ) {
+
+			if ( this.isPlaceholder( node ) ) {
+				this.recordPlaceholderNode( node );
+			}
+
+			// we need to find matching children because the mutation observer
+			// only fires for the top element of an added subtree
+			if ( node.querySelectorAll ) {
+				// funky syntax because NodeList walks like an array but doesn't quack like one
```

Comment:
Good suggestion. Fixed in https://github.com/Automattic/wp-calypso/commit/c17b1fcc529bbe615131736f0bac4fd7ddbe0d86